rskj-core/src/main/java/co/rsk/logfilter/BlocksBloomStore.java

```diff
@@ -40,22 +44,25 @@ public BlocksBloomStore(int noBlocks, int noConfirmations, KeyValueDataSource da
         this.dataSource = dataSource;
     }
 
-    public boolean hasBlockNumber(long blockNumber) {
-        if (this.blocksBloom.containsKey(this.firstNumberInRange(blockNumber))) {
-            return true;
-        }
+    public synchronized boolean hasBlockNumber(long blockNumber) {
+        long key = this.firstNumberInRange(blockNumber);
 
-        if (this.dataSource != null && this.dataSource.get(longToKey(blockNumber)) != null) {
-            return true;
-        }
+        return hasBlockNumberInCache(key)
+            || hasBlockNumberInStore(key);
+    }
+
+    private boolean hasBlockNumberInStore(long key) {
+        return this.dataSource != null && this.dataSource.get(longToKey(key)) != null;
```

Comment:
Seems `dataSource` field cannot be `null` (`non-nullable`)? It's being checked for `null` in a few places in this class. But in other places (e.g. in `flush` method) it is not.

Reply:
Done, I just added additional checks